File: android-gif-drawable/src/main/java/pl/droidsonroids/gif/GifIOException.java
package pl.droidsonroids.gif;

import androidx.annotation.NonNull;

import java.io.IOException;

/**
 * Exception encapsulating {@link GifError}s.
 *
 * @author koral--
 */
public class GifIOException extends IOException {
	private static final long serialVersionUID = 13038402904505L;
	/**
	 * Reason which caused an exception
	 */
	@NonNull
	public final GifError reason;

	private final String mErrnoMessage;

	@Override
	public String getMessage() {
		if (mErrnoMessage == null) {
			return reason.getFormattedDescription();
		}
		return reason.getFormattedDescription() + ": " + mErrnoMessage;
	}

	GifIOException(int errorCode, String errnoMessage) {
		reason = GifError.fromCode(errorCode);
		mErrnoMessage = errnoMessage;
	}

	static GifIOException fromCode(final int nativeErrorCode) {
		if (nativeErrorCode == GifError.NO_ERROR.errorCode) {
			return null;
		}
		return new GifIOException(nativeErrorCode, null);
	}
}

File: android-gif-drawable/src/main/java/pl/droidsonroids/gif/GifIOException.kt
package pl.droidsonroids.gif

import java.io.IOException

/**
 * Exception encapsulating [GifError]s.
 *
 * @author koral--
 */
class GifIOException internal constructor(errorCode: Int, private val mErrnoMessage: String?) :
    IOException() {
    /**
     * Reason which caused an exception
     */
    val reason: GifError
    override val message: String
        get() = if (mErrnoMessage == null) {
            reason.formattedDescription
        } else "${reason.formattedDescription}: $mErrnoMessage"

    init {
        reason = GifError.fromCode(errorCode)
    }

    companion object {
        private const val serialVersionUID = 13038402904505L
        fun fromCode(nativeErrorCode: Int?): GifIOException? {
            return if (nativeErrorCode != null && nativeErrorCode == GifError.NO_ERROR.errorCode) {
                null
            } else GifIOException(nativeErrorCode!!, null)
        }
    }
}